### ISSUE TITLE
7.65.3: "Send failure: Broken pipe" error when existing connection has died 

### ISSUE THREAD

**Issue by jeroen** [curl/curl#4246]
__[updated]__ After Debian upgraded to curl 7.65.3, a unit tests from the R bindings that has been stable for the last years, started failing intermittently with this error: `Send failure: Broken pipe`.

The bug is that curl does not always detect a died connection.

The unit test performs a series of requests to a local http server, each in a separate easy handle but using one and the same multi-handle that keep track of existing connections. However in between each request the http server is restarted, which obviously kills the existing connections.

## Correct behavior (7.65.1)

Before each request curl correctly detects that the existing connection is dead:

```
* Connection #2 to host 127.0.0.1 left intact
* Found bundle for host 127.0.0.1: 0x7f945eda56c0 [serially]
* Connection 2 seems to be dead!
* Closing connection 2
* Hostname 127.0.0.1 was found in DNS cache
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#3)
```

We do this many times and each time curl recognizes when a connection has died and starts a new one, and then succeeds in the request.

## New behavior (7.65.3)

Full example below. Note that it has not noticed that the `Connection #0` has died until after it has tried to send the second request (`Recv failure: Connection reset by peer`). It then recovers with `Connection died, retrying a fresh connect`. 

However when it happens again later at `Re-using existing connection! (#2) with host 127.0.0.1` it seems to fail completely with `Send failure: Broken pipe`.

```
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 26300858
Content-Type: multipart/form-data; boundary=------------------------b48be84cafc652cc

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:16 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #0 to host 127.0.0.1 left intact
* Found bundle for host 127.0.0.1: 0x7fc0015b9f90 [serially]
* Re-using existing connection! (#0) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25193058
Content-Type: multipart/form-data; boundary=------------------------ae32bd27cbda0f89

* Recv failure: Connection reset by peer
* Connection died, retrying a fresh connect
* Closing connection 0
* Issue another request to this URL: 'http://127.0.0.1:9359/'
* Hostname 127.0.0.1 was found in DNS cache
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#1)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25193058
Content-Type: multipart/form-data; boundary=------------------------f1c8bb9f687fb628

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:17 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #1 to host 127.0.0.1 left intact
* Connection 1 seems to be dead!
* Closing connection 1
* Hostname 127.0.0.1 was found in DNS cache
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#2)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25994122
Content-Type: multipart/form-data; boundary=------------------------b732842d3a248b04

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:18 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #2 to host 127.0.0.1 left intact
* Found bundle for host 127.0.0.1: 0x7fc0014471a0 [serially]
* Re-using existing connection! (#2) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#2)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 26011426
Content-Type: multipart/form-data; boundary=------------------------3fcc55e2d2db469e

* Send failure: Broken pipe
* Closing connection 2
```

**Comment by bagder** [comment]
HTTP/1 ?

**Comment by jeroen** [comment]
Yes I'm seeing it on MacOS too with homebrew [curl](https://github.com/homebrew/homebrew-core/blob/master/Formula/curl.rb) (7.65.3 + securetransport, no http/2)

**Comment by jeroen** [comment]
Hmm it also happens for small file uploads:

```
Found bundle for host localhost: 0x7fc4146ae560 [serially]
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: localhost:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 1226
Content-Type: multipart/form-data; boundary=------------------------833cd85c613efd3c

* Send failure: Broken pipe
* Closing connection 0
```

**Comment by bagder** [comment]
Needless to say perhaps, but that's not reported by others so you need to be more specific about how to get that...

**Comment by jeroen** [comment]
It never happens on the first request, but sometimes on the second one or after that (each request has a unique handle but share a single multi handle):

```
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 9359 failed: Connection refused
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: localhost:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 26300858
Content-Type: multipart/form-data; boundary=------------------------86cfe3d4ffb71ebb

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:18:48 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
* Found bundle for host localhost: 0x7faab1f860f0 [serially]
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: localhost:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25193058
Content-Type: multipart/form-data; boundary=------------------------7479d0d540b6c8b6

* Send failure: Broken pipe
* Closing connection 0
```

**Comment by jeroen** [comment]
Sorry for the spam, I was thinking out loud.

I realize now that that the test is somewhat odd because it restarts the local http server in between each request. 

The bug here seems to be that curl not always detects when the server has closed a connection. So in the output below, each request should have `Connection died, retrying a fresh connect`. However instead we see this:

```
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 26300858
Content-Type: multipart/form-data; boundary=------------------------b48be84cafc652cc

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:16 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #0 to host 127.0.0.1 left intact
* Found bundle for host 127.0.0.1: 0x7fc0015b9f90 [serially]
* Re-using existing connection! (#0) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#0)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25193058
Content-Type: multipart/form-data; boundary=------------------------ae32bd27cbda0f89

* Recv failure: Connection reset by peer
* Connection died, retrying a fresh connect
* Closing connection 0
* Issue another request to this URL: 'http://127.0.0.1:9359/'
* Hostname 127.0.0.1 was found in DNS cache
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#1)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25193058
Content-Type: multipart/form-data; boundary=------------------------f1c8bb9f687fb628

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:17 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #1 to host 127.0.0.1 left intact
* Connection 1 seems to be dead!
* Closing connection 1
* Hostname 127.0.0.1 was found in DNS cache
*   Trying 127.0.0.1:9359...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#2)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 25994122
Content-Type: multipart/form-data; boundary=------------------------b732842d3a248b04

* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 20 Aug 2019 14:26:18 GMT
< Content-Type: text/plain
< Content-Length: 0
< 
* Connection #2 to host 127.0.0.1 left intact
* Found bundle for host 127.0.0.1: 0x7fc0014471a0 [serially]
* Re-using existing connection! (#2) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 9359 (#2)
> POST / HTTP/1.1
Host: 127.0.0.1:9359
User-Agent: R (3.6.1 x86_64-apple-darwin15.6.0 x86_64 darwin15.6.0)
Accept: */*
Accept-Encoding: deflate, gzip
Content-Length: 26011426
Content-Type: multipart/form-data; boundary=------------------------3fcc55e2d2db469e

* Send failure: Broken pipe
* Closing connection 2
```

**Comment by jeroen** [comment]
Sorry for the spam, while I was figuring this out. I have updated the topic above and removed some comments with information that turned out to be unrelated.

**Comment by bagder** [comment]
Do you set `CURLOPT_NOSIGNAL` ?

**Comment by jeroen** [comment]
> Do you set CURLOPT_NOSIGNAL ?

no

**Comment by bagder** [comment]
"Broken pipe" means EPIPE, which should make `Curl_send_plain` return `CURLE_SEND_ERROR`.

The logic that is *supposed* to catch this for when a server has closed down in the mean-time is here:

https://github.com/curl/curl/blob/48f589893d483684bac3ff2aebc8ff64f14442af/lib/multi.c#L1637-L1638

I don't think logic has changed recently, at least not on purpose. There's of course a race situation here so that if you close the server slightly after the request goes out there, curl may of course still fail to send the request but then that's also what's to be expected I think.

**Comment by jeroen** [comment]
> There's of course a race situation here so that if you close the server slightly after the request goes out 

This is not the case. The unit tests first restarts the server, and then performs the next request.

**Comment by jeroen** [comment]
```
755083d00deb167667882e775b0885da0e63d034 is the first bad commit
commit 755083d00deb167667882e775b0885da0e63d034
Author: Daniel Stenberg <daniel@haxx.se>
Date:   Sun Jun 16 23:25:38 2019 +0200

    conn_maxage: move the check to prune_dead_connections()

    ... and avoid the locking issue.

    Reported-by: Kunal Ekawde
    Fixes #4029
    Closes #4032
```

**Comment by jeroen** [comment]
It could very well be an interaction with my unusual testing environment (which restarts the httpd and performs a request immediately after). I fixed the test by using `CURLOPT_FORBID_REUSE` so it is not super urgent for me that this is fixed (but I'm quite sure there is a little regression here).

**Comment by a7ypically** [comment]
I'm having a somewhat similar issue.
I'm using https-dns-proxy which in turn uses curl to resolve dns queries over https. Running on x86 with curl 7.65.3, I would get period of times where the proxy would not respond. Looking at the logs I can see that curl returns: CURLINFO_RESPONSE_CODE: 0
It would do so for about a couple of minutes with any request in that period of time. Then it seems that curl recovers until the next issue.
Once I downgraded libcurl to 7.65.1 I can't reproduce this issue, so this might be the same dead connection detection issue reported here.

**Comment by pbomta** [comment]
I have found something in the code that could be the problem.
In #4032 the age check on connections was moved to `prune_dead_connections`. (File: `url.c`) But in doing so, the check for a dead connection (`extract_if_dead`) in `ConnectionExists` was also removed. Now look at `create_conn`. It calls `prune_dead_connections` first, which may or may not do anything, dependent on the timing. It makes certain there is at least 1 second between prunings. Let's say it did nothing. Now, the control flow can go to `ConnectionExists` and there it can try to reuse a dead connection.

Some of our tests started failing after upgrading from Curl 7.65.1 to 7.65.3. I have not been able to isolate the problem yet, so regrettably, I have no quick & easy reproduction.

@jeroen Perhaps you can disable CURLOPT_FORBID_REUSE and add the dead check? This is the patch to apply:
```
--- a/Framework/ThirdParty/Curl/Implementation/url.c 
+++ b/Framework/ThirdParty/Curl/Implementation/url.c 
@@ -1077,6 +1077,11 @@ ConnectionExists(struct Curl_easy *data, 
         /* connect-only connections will not be reused */ 
         continue; 
  
+      if(extract_if_dead(check, data)) { 
+        Curl_disconnect(data, check, TRUE); 
+        continue; 
+      } 
+ 
       multiplexed = CONN_INUSE(check) && 
         (bundle->multiuse == BUNDLE_MULTIPLEX);
```

**Comment by xquery** [comment]
not able to repro (with latest master) though I also cannot claim I understand test env completely ... considering @a7ypically last comment about things working in openwrt perhaps there is no problem here ... any chance of getting a check using latest release ?

**Comment by a7ypically** [comment]
I can't reproduce with the latest either.

**Comment by bagder** [comment]
@jeroen can you still reproduce this with the latest libcurl?

**Comment by jeroen** [comment]
Sadly, I tried to reconstruct my test environment, but I can no longer reproduce this problem, not even with curl 7.65.3 (the version that I originally found this problem in). 

There must be some external factor that has changed in my environment which was triggering this bug. Perhaps `EPIPE` has changed in the new MacOS (though we saw the bug both in Debian and Mac)....

I'm sorry I cannot give a better understanding, I guess we can close the issue. This bug was difficult to reproduce to begin with, only appearing under unusual circumstances. Hopefully the problem is resolved, whatever it was 😄
